### PR TITLE
[REEF-676] Made *Task.ActiveContext immutable

### PR DESCRIPTION
**Pull request by markusweimer**
* Made `ActiveContext` immutable in `RunningTask`, `CompletedTask` and
    `SuspendedTask` implementations and interfaces.
  * Deleted the spurious `IRunningTask` in O.A.R.Common. To do this, I
    had to move `EvaluatorException` into the `Driver`.
  * In order to make that new `EvaluatorException` `internal`, I added
    the Bridge as a friend assembly of the Driver.

JIRA:
  [REEF-676](https://issues.apache.org/jira/browse/REEF-676)
  [REEF-668](https://issues.apache.org/jira/browse/REEF-668)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        private readonly string _evaluatorId;
+        private readonly IRunningTask _runningTask;
+
+        internal EvaluatorException(string evaluatorId)
+        {
+            _evaluatorId = evaluatorId;
+            _runningTask = null;
+        }
+
+        internal EvaluatorException(string evaluatorId, string message, Exception cause)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: Use optional parameters instead.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: Not sure what the convention is, but it seems like most other files use block comments.

File: lang/cs/Org.Apache.REEF.Driver/Task/IRunningTask.cs [hunk 3]
```diff
@@ -31,13 +31,7 @@ public interface IRunningTask : IIdentifiable, IDisposable
         /// <summary>
         /// the context the task is running on.
         /// </summary>
-        IActiveContext ActiveContext { get; set; }
-
-        /// <summary>
-        /// Sends the message to the running task.
-        /// </summary>
-        /// <param name="message"></param>
-        void OnNext(byte[] message);
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Task/IRunningTask.cs]: Why is this removed?

File: lang/cs/Org.Apache.REEF.Driver/Task/ICompletedTask.cs [hunk 4]
```diff
@@ -24,6 +24,9 @@ namespace Org.Apache.REEF.Driver.Task
 {
     public interface ICompletedTask : IMessage, IIdentifiable
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Task/ICompletedTask.cs]: Could you add documentation for this interface as well? Thanks.

**Comment by afchung** [comment]
Took an initial pass. Only left some minor comments.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: I believe this style is the result of configuring ReSharper to automatically insert a license into each file.

I don't think we have a convention. Should we have one?

File: lang/cs/Org.Apache.REEF.Driver/Task/IRunningTask.cs (hunk 3 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Driver/Task/IRunningTask.cs]: It never was implemented and has the exact same meaning as `Send`, which was implemented.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs (hunk 2 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: I think we should be consistent.

**Comment by afchung** [comment]
@markusweimer Seems like there is a conflict. Could you please resolve? Thanks!

**Comment by afchung** [comment]
@markusweimer Why was this PR closed?

**Comment by markusweimer** [comment]
Sorry, that was by accident. Will resolve conflicts and ping again.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: done.

File: lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Driver/Evaluator/EvaluatorException.cs]: Good call. I copied the code from the old instance of this type. I fixed it to use default parameters instead.

**Comment by markusweimer** [comment]
@afchung I've addressed your feedback and rebased onto current `master`. Time for another look.

**Comment by afchung** [comment]
Looks good. I will test and merge.
